src/main/java/org/primefaces/component/datalist/DataList.java

```diff
@@ -195,4 +196,42 @@ protected void processChildren(FacesContext context, PhaseId phaseId) {
         }
     }
 
+    @Override
+    public boolean isMultiViewState() {
+        return super.isMultiViewState();
+    }
+
+    public void restoreListState() {
+        ListState ls = getListState(false);
+        if (ls != null) {
+            if (isPaginator()) {
+                setFirst(ls.getFirst());
+                int rows = (ls.getRows() == 0) ? getRows() : ls.getRows();
+                setRows(rows);
+            }
+        }
+    }
+
+    public ListState getListState(boolean create) {
+        FacesContext fc = getFacesContext();
+        Map<String, Object> sessionMap = fc.getExternalContext().getSessionMap();
+        Map<String, ListState> dlState = (Map) sessionMap.get(Constants.DATALIST_STATE);
+        String viewId = fc.getViewRoot().getViewId().replaceFirst("^/*", "");
+        String stateKey = viewId + "_" + getClientId(fc);
+        ListState ls;
```

Comment:
Write `ListState ls = dlState.get(stateKey);` instead, before the condition (e.g remove that cumbersome declaration `ListState ls;`)